### DISCUSSION TITLE
Terragrunt parsing support

### DISCUSSION THREAD

**Discussion by zischknall** [suzuki-shunsuke/tfcmt#843] [Deprecated (Feature Request)] [answered]
### Feature Overview

[Terragrunt](https://terragrunt.gruntwork.io/) is a Terraform wrapper to handle DRY configuration of multiple environments managed through Terraform. 

Terragrunt runs multiple modules one after another and has next to no output of it's own, it would be nice to also handle the outputs of those as though they were one Terraform `plan` output.

### Why is the feature needed?

Currently runs of `terragrunt plan` can cause the parser to crash with an index out of bounds if it contains blocks multiple times. For example when the output contains two change blocks the parser crashes.

### Does the feature include Breaking Changes?

By always looking for multiple blocks Terragrunt could be supported at the same time as continuing support for Terraform.
I currently have a fork with what I believe to be working Terragrunt plan parsing (no apply as of yet).
The only breaking change would be that I changed the formatting of the resulting output blocks slightly, while I might argue that it is know formatted in a more controlled manner.

### Example Code

No small code example as it would change the internal parsing behavior.

**Comment by suzuki-shunsuke** [comment]
Thank you for your proposal.
I'm not familiar with Terragrunt at all, so first I'll take a look at it and reproduce the issue to understand the issue.

**Reply by suzuki-shunsuke** [reply]
I could reproduce the issue.

```console
$ terraform version
Terraform v1.5.3
on darwin_arm64

$ tfcmt version
tfcmt version 4.4.2 (49d366319ec5d3170ede4076f3a4c79d1f035638)

$ terragrunt --version
terragrunt version v0.48.1
```

```console
$ tree
.
├── frontend-app
│   ├── main.tf
│   └── terragrunt.hcl
├── mysql
│   ├── main.tf
│   └── terragrunt.hcl
└── terragrunt.hcl

3 directories, 5 files
```

terragrunt.hcl

```hcl
# empty
```

frontend-app/main.tf, mysql/main.tf

```tf
resource "null_resource" "foo" {}
```

`terragrunt run-all plan` succeeded.

https://terragrunt.gruntwork.io/docs/features/execute-terraform-commands-on-multiple-modules-at-once/#the-run-all-command

<details>

```console
$ terragrunt run-all plan
INFO[0000] The stack at /Users/shunsukesuzuki/Documents/test/tfcmt/terragrunt/843/stage will be processed in the following order for command plan:
Group 1
- Module /Users/shunsukesuzuki/Documents/test/tfcmt/terragrunt/843/stage/frontend-app
- Module /Users/shunsukesuzuki/Documents/test/tfcmt/terragrunt/843/stage/mysql
 

Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:

  # null_resource.foo will be created
  + resource "null_resource" "foo" {
      + id = (known after apply)
    }

Plan: 1 to add, 0 to change, 0 to destroy.

─────────────────────────────────────────────────────────────────────────────

Note: You didn't use the -out option to save this plan, so Terraform can't
guarantee to take exactly these actions if you run "terraform apply" now.
  + create

Terraform will perform the following actions:

  # null_resource.foo will be created
  + resource "null_resource" "foo" {
      + id = (known after apply)
    }

Plan: 1 to add, 0 to change, 0 to destroy.

─────────────────────────────────────────────────────────────────────────────

Note: You didn't use the -out option to save this plan, so Terraform can't
guarantee to take exactly these actions if you run "terraform apply" now.
```

</details>

`tfcmt plan -- terragrunt run-all plan` failed.

```console
$ tfcmt plan -- terragrunt run-all plan
time=2023-07-15T09:37:14+09:00 level=info msg=The stack at /Users/shunsukesuzuki/Documents/test/tfcmt/terragrunt/843/stage will be processed in the following order for command plan:
Group 1
- Module /Users/shunsukesuzuki/Documents/test/tfcmt/terragrunt/843/stage/frontend-app
- Module /Users/shunsukesuzuki/Documents/test/tfcmt/terragrunt/843/stage/mysql



Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # null_resource.foo will be created
  + resource "null_resource" "foo" {
      + id = (known after apply)
    }

Plan: 1 to add, 0 to change, 0 to destroy.

Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # null_resource.foo will be created
  + resource "null_resource" "foo" {
      + id = (known after apply)
    }

Plan: 1 to add, 0 to change, 0 to destroy.

─────────────────────────────────────────────────────────────────────────────

Note: You didn't use the -out option to save this plan, so Terraform can't
guarantee to take exactly these actions if you run "terraform apply" now.

─────────────────────────────────────────────────────────────────────────────

Note: You didn't use the -out option to save this plan, so Terraform can't
guarantee to take exactly these actions if you run "terraform apply" now.
panic: runtime error: slice bounds out of range [25:19]

goroutine 1 [running]:
github.com/suzuki-shunsuke/tfcmt/pkg/terraform.(*PlanParser).Parse(0x140000b8370, {0x14000310000, 0x72d})
	/home/runner/work/tfcmt/tfcmt/pkg/terraform/parser.go:183 +0x1130
github.com/suzuki-shunsuke/tfcmt/pkg/notifier/github.(*NotifyService).Plan(0x140000aa198, {0x101100f40, 0x14000036e80}, 0x14000312000)
	/home/runner/work/tfcmt/tfcmt/pkg/notifier/github/plan.go:24 +0xb4
github.com/suzuki-shunsuke/tfcmt/pkg/controller.(*Controller).Plan(0x1400014f710, {0x101100f40, 0x14000036e80}, {{0x16f362f2e, 0xa}, {0x140000a2d20, 0x2, 0x2}})
	/home/runner/work/tfcmt/tfcmt/pkg/controller/plan.go:49 +0xa2c
github.com/suzuki-shunsuke/tfcmt/pkg/cli.cmdPlan(0x140000a8400)
	/home/runner/work/tfcmt/tfcmt/pkg/cli/plan.go:36 +0x2a4
github.com/urfave/cli/v2.(*Command).Run(0x140000a0000, 0x140000a8400, {0x140000b8050, 0x5, 0x5})
	/home/runner/go/pkg/mod/github.com/urfave/cli/v2@v2.25.7/command.go:274 +0x73c
github.com/urfave/cli/v2.(*Command).Run(0x140000a0580, 0x140000a8040, {0x1400001e1e0, 0x6, 0x6})
	/home/runner/go/pkg/mod/github.com/urfave/cli/v2@v2.25.7/command.go:267 +0x948
github.com/urfave/cli/v2.(*App).RunContext(0x14000094000, {0x101100f40?, 0x14000036e80}, {0x1400001e1e0, 0x6, 0x6})
	/home/runner/go/pkg/mod/github.com/urfave/cli/v2@v2.25.7/app.go:332 +0x568
main.core()
	/home/runner/work/tfcmt/tfcmt/cmd/tfcmt/main.go:30 +0x120
main.main()
	/home/runner/work/tfcmt/tfcmt/cmd/tfcmt/main.go:19 +0x1c
```

**Reply by suzuki-shunsuke** [reply]
📝 I understood the cause of the panic.
`startChangeOutput` is the last line of `Terraform will perform the following actions:`,
but `endChangeOutput` is the first line starting with `Plan:`.
So the index of `startChangeOutput` is larger than `endChangeOutput`, so `lines[startChangeOutput:endChangeOutput]` gets a panic.

https://github.com/suzuki-shunsuke/tfcmt/blob/a2dd26e718a7aa8fe7321b99e65773c560082a3c/pkg/terraform/parser.go#L182-L184

https://github.com/suzuki-shunsuke/tfcmt/blob/a2dd26e718a7aa8fe7321b99e65773c560082a3c/pkg/terraform/parser.go#L133-L135

https://github.com/suzuki-shunsuke/tfcmt/blob/a2dd26e718a7aa8fe7321b99e65773c560082a3c/pkg/terraform/parser.go#L136-L138

**Reply by zischknall** [reply]
I think there are three ways to go about this:
1. Ignore it and say that terragrunt is not supported (would be totally fine)
2. Make sure it does not crash but don't change the parsing. Causing only one change block to be commented in a terragrunt run (so somewhat partial support)
3. Change the parsing so it can handle both Terraform and Terragrunt. Officially supporting both

I think all of the above are totally fine and it's ultimately up to you.

**Comment by suzuki-shunsuke** [comment]
What's the desirable hahaviour of this case? 🤔

1. Put together all modules' result in one comment
2. Post comments per module

**Reply by suzuki-shunsuke** [reply]
I prefer `2`, but ideally we can choose both by configuration file and command line argument.

**Reply by suzuki-shunsuke** [reply]
## Idea 1

```console
$ tfcmt -terragrunt-run-all -split-comment plan 
```

tfcmt.yaml

```yaml
terragrunt:
  run-all:
    split_comment: true
```

## Idea 2

```console
$ tfcmt -terragrunt-run-all -one-comment plan 
```

tfcmt.yaml

```yaml
terragrunt:
  run-all:
    one_comment: true
```

**Comment by suzuki-shunsuke** [comment]
📝 We need to get the list of modules from the following output.

```
Group 1
- Module /Users/shunsukesuzuki/Documents/test/tfcmt/terragrunt/843/stage/frontend-app
- Module /Users/shunsukesuzuki/Documents/test/tfcmt/terragrunt/843/stage/mysql
```

https://github.com/gruntwork-io/terragrunt/blob/eda265f2dc6b64dba46f1f5e618cd9ec28cd8c97/configstack/stack.go#L34-L52

**Reply by suzuki-shunsuke** [reply]
We need to the relative path from the current directory.

`/Users/shunsukesuzuki/Documents/test/tfcmt/terragrunt/843/stage/frontend-app` => `frontend-app`

Sometimes we need to convert the directory file path to a target.

**Comment by zischknall** [comment]
Sorry at first only saw your first answer. Never worked with the discussions feature before and the mobile UI is a bit clunky. From my point of view at first I was fine with just concatenating all of the modules output blocks.
But I like the idea of showing it per module. But if I remember correctly Terragrunt shows at the beginning what modules it applies but in the following output it is difficult to discern which output belongs to which module.

**Comment by suzuki-shunsuke** [comment, marked as answer]
I found that we can post the result of terragrunt with tfcmt by terragrunt [--terragrunt-tfpath](https://terragrunt.gruntwork.io/docs/reference/cli-options/#terragrunt-tfpath) option.

1. Create a wrapper script of `terraform` and make it executable

```console
$ vi tfwrapper.sh
$ chmod a+x tfwrapper.sh
```

tfwrapper.sh

```sh
#!/bin/bash

set -euo pipefail

command=$1

base_dir=$(git rev-parse --show-toplevel) # Please fix if necessary
target=${PWD#"$base_dir"/}

if [ "$command" == "plan" ]; then
  tfcmt -var "target:${target}" plan -- terraform "$@"
elif [ "$command" == "apply" ]; then
  tfcmt -var "target:${target}" apply -- terraform "$@"
else
  terraform "$@"
fi
```

2. Run `terragrunt run-all` with `--terragrunt-tfpath`

```console
$ terragrunt run-all plan --terragrunt-tfpath "<absolute path of tfwrapper.sh>"
```

Then the result of `terraform plan` is posted per module.

**Reply by zischknall** [reply]
Nice idea. I will try it in the following days and get back to you.

**Reply by suzuki-shunsuke** [reply]
Thank you. I'm looking forward to your feedback.
I added a document.
https://suzuki-shunsuke.github.io/tfcmt/terragrunt-run-all/

**Reply by zischknall** [reply]
Mhm, that does not really work out for me.
Let me explain we have 4 folders on the root level, one for each environment ci, qa, pp, prod.
The modules in each of these folders are pretty much the same.
Now if someone opens a pull request I want to test all four environments and add a plan comment per environment.
I do this by starting a workflow run that runs with a matrix one for each environment.
With the wrapper script I now get one comment for each module in the environments, which makes it hard to recognize in which environment this is.

Also some modules throw errors, I don't know why yet.
```
time=2023-07-17T07:54:39Z level=info msg=time="2023-07-17T07:53:58Z" level=error msg="add a label" error="POST https://api.github.com/repos/xxx/xxx/issues/24/labels: 422 Validation Failed [{Resource:Label Field:name Code:invalid Message:}]" label=/home/runner/work/xxx/xxx/projects/ci/buckets/pipeline/.terragrunt-cache/ZQKh0uHBw_1pS35ve2ew8rzhJgs/ufzFrCkDUnPpPM2hdbErkJoUWvU/no-changes program=tfcmt

```

**Reply by suzuki-shunsuke** [reply]
The error occurred because the pull request label was too long.
IIRC, the maximum length of GitHub label is 50.

`label=/home/runner/work/xxx/xxx/projects/ci/buckets/pipeline/.terragrunt-cache/ZQKh0uHBw_1pS35ve2ew8rzhJgs/ufzFrCkDUnPpPM2hdbErkJoUWvU/no-changes`

It seems the target argument is wrong.

```sh
base_dir=$(git rev-parse --show-toplevel) # Please fix if necessary
target=${PWD#"$base_dir"/}
```

I'm not sure about your environment, so please fix the above code for your environment.
Maybe I can help you if you tell me your environment.

> With the wrapper script I now get one comment for each module in the environments, which makes it hard to recognize in which environment this is.

You can distinguish comments by comment header.

https://suzuki-shunsuke.github.io/tfcmt/getting-started#monorepo-support-target-variable

e.g. `target=foo`

![image](https://user-images.githubusercontent.com/13323303/111018399-ea8bea00-83fb-11eb-8efe-205ab8c996b7.png)

**Reply by zischknall** [reply]
This is what our setup looks like:
<img width="327" alt="image" src="https://github.com/suzuki-shunsuke/tfcmt/assets/28953242/88389a0d-93cf-423b-aabf-a432210ca74c">

I guess the target variable should just be one of CI, QA, PP or PROD?

**Reply by suzuki-shunsuke** [reply]
`target` is used to identify working directories.

Where is your `terragrount.hcl` and Terraform Configuration files?
For example, if `terragrunt.hcl` and Terraform Configuration files are located in the directory `projects/qa/europe-west1/alloydb`,
`target` should be `projects/qa/europe-west1/alloydb`. Or you can also omit the prefix `projects/`.

```
label=/home/runner/work/xxx/xxx/projects/ci/buckets/pipeline/.terragrunt-cache/ZQKh0uHBw_1pS35ve2ew8rzhJgs/ufzFrCkDUnPpPM2hdbErkJoUWvU/no-changes
```

This seems that terragrunt ran the wrapper script in the cache directory `/home/runner/work/xxx/xxx/projects/ci/buckets/pipeline/.terragrunt-cache/ZQKh0uHBw_1pS35ve2ew8rzhJgs/ufzFrCkDUnPpPM2hdbErkJoUWvU`. 🤔 
I'm not sure how terragrunt works.

**Reply by suzuki-shunsuke** [reply]
Could you output the debug log in tfwrapper.sh?

```sh
base_dir=$(git rev-parse --show-toplevel) # Please fix if necessary
target=${PWD#"$base_dir"/}
echo "[DEBUG] base_dir: $base_dir" >&2
echo "[DEBUG] PWD: $PWD" >&2
echo "[DEBUG] target: $target" >&2
```

**Reply by zischknall** [reply]
Terragrunt is run with `run-all` from each project folder so, once for `qa` and once for `pp` and so on.
Ideally I would want to have one comment on the Pull Request for each of those environments. Because we are running a plan on all four environments for each module. It would get quite overwhelming if there was a comment for each module.

Also I will try the DEBUG outputs and get back to you with that as well.

**Reply by zischknall** [reply]
Here is some of the DEBUG output:
```
[DEBUG] PWD: /home/runner/work/xxx/xxx/projects/ci/buckets/terraform-state/.terragrunt-cache/iH1mOlqPdLCgP0MJTfatAPSEYfs/ufzFrCkDUnPpPM2hdbErkJoUWvU
[DEBUG] target: /home/runner/work/xxx/xxx/projects/ci/buckets/terraform-state/.terragrunt-cache/iH1mOlqPdLCgP0MJTfatAPSEYfs/ufzFrCkDUnPpPM2hdbErkJoUWvU
[DEBUG] base_dir: /home/runner/work/xxx/xxx/projects/ci/wif-gh/.terragrunt-cache/BcrPECXlqn7U-5Fotql61tZ7Ejo/Z980WgVAL1mpXzVQ-9hVUsPNkOs
[DEBUG] PWD: /home/runner/work/xxx/xxx/projects/ci/wif-gh/.terragrunt-cache/BcrPECXlqn7U-5Fotql61tZ7Ejo/Z980WgVAL1mpXzVQ-9hVUsPNkOs/wif-gh
[DEBUG] target: wif-gh
[DEBUG] base_dir: /home/runner/work/xxx/xxx/projects/ci/project-cfg/.terragrunt-cache/-36JXAhoYXLANqQK-lAYxNcoalw/Nd3sy7QfqxLOiPv0uS3AyL_BjyM
[DEBUG] PWD: /home/runner/work/xxx/xxx/projects/ci/project-cfg/.terragrunt-cache/-36JXAhoYXLANqQK-lAYxNcoalw/Nd3sy7QfqxLOiPv0uS3AyL_BjyM
[DEBUG] target: /home/runner/work/xxx/xxx/projects/ci/project-cfg/.terragrunt-cache/-36JXAhoYXLANqQK-lAYxNcoalw/Nd3sy7QfqxLOiPv0uS3AyL_BjyM
[DEBUG] base_dir: /home/runner/work/xxx/xxx/projects/ci/project-cfg/.terragrunt-cache/-36JXAhoYXLANqQK-lAYxNcoalw/Nd3sy7QfqxLOiPv0uS3AyL_BjyM
[DEBUG] PWD: /home/runner/work/xxx/xxx/projects/ci/project-cfg/.terragrunt-cache/-36JXAhoYXLANqQK-lAYxNcoalw/Nd3sy7QfqxLOiPv0uS3AyL_BjyM
[DEBUG] target: /home/runner/work/xxx/xxx/projects/ci/project-cfg/.terragrunt-cache/-36JXAhoYXLANqQK-lAYxNcoalw/Nd3sy7QfqxLOiPv0uS3AyL_BjyM
```

**Reply by suzuki-shunsuke** [reply]
> Terragrunt is run with run-all from each project folder so, once for qa and once for pp and so on.

Yeah, that is what I expect, so I can't understand why `target` was `/home/runner/work/xxx/xxx/projects/ci/buckets/pipeline/.terragrunt-cache/ZQKh0uHBw_1pS35ve2ew8rzhJgs/ufzFrCkDUnPpPM2hdbErkJoUWvU`.

> Ideally I would want to have one comment on the Pull Request for each of those environments.

I understand your needs.
But tfcmt doesn't support it for now.

My concern is that if results are put together in one comment, the comment is so large that it is difficult to read and the length may reach  the limitation of the comment length.

**Reply by suzuki-shunsuke** [reply]
The debug log looks weird.

```console
$ pbpaste | grep PWD   
[DEBUG] PWD: /home/runner/work/xxx/xxx/projects/ci/buckets/terraform-state/.terragrunt-cache/iH1mOlqPdLCgP0MJTfatAPSEYfs/ufzFrCkDUnPpPM2hdbErkJoUWvU
[DEBUG] PWD: /home/runner/work/xxx/xxx/projects/ci/wif-gh/.terragrunt-cache/BcrPECXlqn7U-5Fotql61tZ7Ejo/Z980WgVAL1mpXzVQ-9hVUsPNkOs/wif-gh
[DEBUG] PWD: /home/runner/work/xxx/xxx/projects/ci/project-cfg/.terragrunt-cache/-36JXAhoYXLANqQK-lAYxNcoalw/Nd3sy7QfqxLOiPv0uS3AyL_BjyM
[DEBUG] PWD: /home/runner/work/xxx/xxx/projects/ci/project-cfg/.terragrunt-cache/-36JXAhoYXLANqQK-lAYxNcoalw/Nd3sy7QfqxLOiPv0uS3AyL_BjyM
```

1. The command was run twice in the same directory `/home/runner/work/xxx/xxx/projects/ci/project-cfg/.terragrunt-cache/-36JXAhoYXLANqQK-lAYxNcoalw/Nd3sy7QfqxLOiPv0uS3AyL_BjyM`.

2. `base_dir` was outputted only three times.

```console
$ pbpaste | grep base_dir
[DEBUG] base_dir: /home/runner/work/xxx/xxx/projects/ci/wif-gh/.terragrunt-cache/BcrPECXlqn7U-5Fotql61tZ7Ejo/Z980WgVAL1mpXzVQ-9hVUsPNkOs
[DEBUG] base_dir: /home/runner/work/xxx/xxx/projects/ci/project-cfg/.terragrunt-cache/-36JXAhoYXLANqQK-lAYxNcoalw/Nd3sy7QfqxLOiPv0uS3AyL_BjyM
[DEBUG] base_dir: /home/runner/work/xxx/xxx/projects/ci/project-cfg/.terragrunt-cache/-36JXAhoYXLANqQK-lAYxNcoalw/Nd3sy7QfqxLOiPv0uS3AyL_BjyM
```

**Reply by zischknall** [reply]
I think I will continue to maintain my fork then and improve on it. I have not run into any comment body limits yet. I will deal with those as they appear.
